File: 백준/Gold/2406. 안정적인 네트워크/안정적인 네트워크.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();
    static StringTokenizer st;
    static int[] parent, rank;
    static int sum = 0, cnt = 0;

    static class Node implements Comparable<Node> {
        int u, v, d;

        public Node(int u, int v, int d) {
            this.u = u;
            this.v = v;
            this.d = d;
        }

        @Override
        public int compareTo(Node other) {
            return this.d - other.d;
        }
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        PriorityQueue<Node> pq = new PriorityQueue<>();

        parent = new int[n];
        rank = new int[n];

        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            union(Integer.parseInt(st.nextToken()) - 1, Integer.parseInt(st.nextToken()) - 1);
        }

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                int num = Integer.parseInt(st.nextToken());
                if (i == 0 || j == 0 || i == j) continue;
                pq.add(new Node(i, j, num));
            }
        }

        while (!pq.isEmpty()) {
            Node c = pq.poll();

            if (find(c.u) != find(c.v)) {
                union(c.u, c.v);
                sum += c.d;
                cnt++;
                sb.append((c.u + 1)).append(" ").append((c.v + 1)).append("\n");
            }
        }

        System.out.println(sum + " " + cnt);
        System.out.print(sb);
    }

    static int find(int node) {
        if (parent[node] != node) {
            parent[node] = find(parent[node]);
        }
        return parent[node];
    }

    static void union(int node1, int node2) {
        int root1 = find(node1);
        int root2 = find(node2);

        if (root1 == root2) return;

        if (rank[root1] > rank[root2]) {
            parent[root2] = root1;
        } else if (rank[root1] < rank[root2]) {
            parent[root1] = root2;
        } else {
            parent[root2] = root1;
            rank[root1]++;
        }
    }
}

File: 백준/Gold/2406. 안정적인 네트워크/안정적인 네트워크.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

data class Node(val u: Int, val v: Int, val d: Int)

val sb = StringBuilder()
var parent = intArrayOf()
var rank = intArrayOf()
var sum = 0
var result = 0
var cnt = 0

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()

    val pq = PriorityQueue<Node>(compareBy { it.d })

    parent = IntArray(n) { it }
    rank = IntArray(n)

    repeat(m) {
        val edge = StringTokenizer(br.readLine())
        val a = edge.nextToken().toInt() - 1
        val b = edge.nextToken().toInt() - 1

        union(a, b, 0)
    }

    repeat(n) { i ->
        val st2 = StringTokenizer(br.readLine())

        for (j in 0 until n) {
            val num = st2.nextToken().toInt()
            if (i == 0 || j == 0 || i == j) continue

            pq.add(Node(i, j, num))
        }
    }

    while (pq.isNotEmpty()) {
        val c = pq.poll()

        union(c.u, c.v, c.d)
        if (cnt == n - 2) break
    }

    println("$sum $result")
    print(sb)
}

fun find(node: Int): Int {
    if (parent[node] != node) {
        parent[node] = find(parent[node])
    }
    return parent[node]
}

fun union(node1: Int, node2: Int, num: Int) {
    val root1 = find(node1)
    val root2 = find(node2)

    if (root1 == root2) return

    if (rank[root1] > rank[root2]) {
        parent[root2] = root1
    } else if (rank[root1] < rank[root2]) {
        parent[root1] = root2
    } else {
        parent[root2] = root1
        rank[root1]++
    }

    sum += num
    cnt++

    if (num != 0) {
        result++
        sb.appendLine("${node1 + 1} ${node2 + 1}")
    }
}